spec/requests/openid_connect_authorize_spec.rb

```diff
@@ -28,4 +28,55 @@
     expect(response).
       to redirect_to login_two_factor_path(otp_delivery_preference: 'sms', reauthn: false)
   end
+
+  it 'prompts the user to 2FA if prompt is login' do
+    client_id = 'urn:gov:gsa:openidconnect:sp:server'
+    state = SecureRandom.hex
+    nonce = SecureRandom.hex
+
+    post new_user_session_path, params: { user: { email: user.email, password: user.password } }
+    follow_redirect!
+    get(
+      openid_connect_authorize_path,
+      params: {
+        client_id: client_id,
+        response_type: 'code',
+        acr_values: Saml::Idp::Constants::LOA3_AUTHN_CONTEXT_CLASSREF,
+        scope: 'openid email profile:name social_security_number',
+        redirect_uri: 'http://localhost:7654/auth/result',
+        state: state,
+        prompt: 'login',
+        nonce: nonce,
+      },
+      headers: { 'Accept' => '*/*' }
+    )
+    follow_redirect!
+    expect(response).
+      to redirect_to login_two_factor_path(otp_delivery_preference: 'sms', reauthn: false)
+  end
+
+  it 'prompts the user to 2FA if prompt is not given' do
+    client_id = 'urn:gov:gsa:openidconnect:sp:server'
+    state = SecureRandom.hex
+    nonce = SecureRandom.hex
+
+    post new_user_session_path, params: { user: { email: user.email, password: user.password } }
+    follow_redirect!
+    get(
+      openid_connect_authorize_path,
+      params: {
+        client_id: client_id,
+        response_type: 'code',
+        acr_values: Saml::Idp::Constants::LOA3_AUTHN_CONTEXT_CLASSREF,
+        scope: 'openid email profile:name social_security_number',
+        redirect_uri: 'http://localhost:7654/auth/result',
+        state: state,
+        nonce: nonce,
+      },
+      headers: { 'Accept' => '*/*' }
+    )
+    follow_redirect!
+    expect(response).
+      to redirect_to login_two_factor_path(otp_delivery_preference: 'sms', reauthn: false)
+  end
```

Comment:
What scenario(s) are we implementing this for? Is it when the user is signed in via email and password only, and then tries to login from the SP? Or is it when the user is fully signed in (including 2FA), then tries to sign in again from the SP? Or both? If both, can we add specs for the fully signed in scenario please?

Reply:
Sorry, didn't realize this is a WIP. I will offer one last suggestion for the tests and that is to simulate as closely as possible real life scenarios, such as first making a request from an SP, then signing in the user, then trying to make another request from the SP with a different prompt. Maybe feature specs would be better to test this. See these specs for examples: https://github.com/18F/identity-idp/blob/master/spec/features/openid_connect/openid_connect_spec.rb

Reply:
yes definitely a feature spec